### DISCUSSION TITLE
main.py update gracefully?

### DISCUSSION THREAD

**Discussion by roozeboos** [micropython/micropython#13093] [RP2040 / Pico]
Hi All,
Ive got a W5100-evb-pico running. Developing with thonny. All is well. 
When I copy main.py to the module is runs fine.

Only when I want to update the main.py I cannot get REPL to work. at button press it exits and shows :>> in thonny (as expected)
But then a second later it gives: Connection lost -- EOF Use Stop/Restart to reconnect. Process ended with exit code 1. 
and the comport is closed again!
The only way to recover is nuke it and reprogramm the module.
Is there a way to avoid this full reprogramm??

MicroPython v1.21.0 on 2023-10-06 and W5100S_EVB_PICO-20231005-v1.21.0.uf2 is used.

regards,
Chris from holland

**Comment by robert-hh** [comment]
Skip Thonny for that task, use mpremote.
Besides that: put your code into another file of your choice and name, like my_app.py, and call it with `include my_app` from main.py. You can as well add other measures like testing the value of a pin before doing so, allowing to a  skip execution of you app if it should not.

**Reply by roozeboos** [reply]
Hi Robert, thanks very much for your reply! I will  try mpremote.
I have just tried a simple led blink main.py. and that exists fine.
So it must also be something in my code.
Wehen debugging I also need to press the run button to exit, stop doenst work, I guess that is an indication?

**Comment by roozeboos** [comment]
it looks like the socket accept() messes thing up. When I take that out, it exits normally.
How can I solve that?

**Comment by roozeboos** [comment]
surely something todo with settimeout to a non zero value..

**Comment by roozeboos** [comment]
timeout  to zero solved it.
So I guess thonny doesnt like the blocking of the socket

**Comment by roozeboos** [comment]
I finally installed and learned about mpremote. Best advise, that works a charm!
Thanks again Robert,
